New validator: mutually exclusive tags

mtmail opened the pull request:
Closes #8306, Closes #7589

About 20.000 places currently have `name` and `noname` set. https://overpass-turbo.eu/s/1Eyn (map screenshot and discussion at https://community.openstreetmap.org/t/places-with-name-but-also-noname-yes/106858)

I chose 'mutually exclusive' but 'incompatible' or 'conflicting' could also work. I'm open to wording suggestions.

Future versions might need to handle values, too. For example `["wheelchair=yes", "ladder=yes"]`. Or patterns `["noname", "name:*"]`

In the future I'd like to see the list of pairs maintained by the community in https://github.com/openstreetmap/id-tagging-schema/ like the list of deprecated tags.

Thank you @tyrasd for the iD development workshop at SotM Antwerp.

![image](https://github.com/openstreetmap/iD/assets/3727288/9fa2e97a-e23f-42b9-8547-45600fab96b7)

mtmail committed:
new validator: mutually exclusive tags

1ec5 commented on modules/validations/mutually_exclusive_tags.js:
```diff
@@ -0,0 +1,46 @@
+export function validationMutuallyExclusiveTags(/* context */) {
+    const type = 'mutually_exclusive_tags';
+
+    // https://wiki.openstreetmap.org/wiki/Special:WhatLinksHere/Property:P44
```
Good idea to consult this property. 👍

1ec5 commented on modules/validations/mutually_exclusive_tags.js:
```diff
@@ -0,0 +1,46 @@
+
+    // https://wiki.openstreetmap.org/wiki/Special:WhatLinksHere/Property:P44
+    const tagPairs = [
+        ['noname', 'name'],
```
#6411 added a check for `name` matching `not:name`. It isn’t quite the same thing as checking whether two keys are merely present, but I wonder if it should be integrated into this validator rule somehow:

https://github.com/openstreetmap/iD/blob/bb17d5c24eb6c58724b8be83c646bfe0aa9cf3ad/modules/validations/suspicious_name.js#L154-L170

In that issue, we considered it important to add a field for `not:name` so that the user wouldn’t have to dig through the raw tags to resolve the issue. I don’t think any of the negative keys here have fields yet. Maybe it wouldn’t block this PR at all, but we should consider adding fields to id-tagging-schema separately.

mtmail replied:
So basically whenever a "tagname" and "not:tagname" is found and both have the same value then validation should fail? I think it makes sense to move that logic into this new validator.

1ec5 replied:
Yes, that’s what I was thinking. However, another difference between this validator rule and the existing `not:name` rule is that the latter validates `name:*` subkeys against each value in `not:name`. We could justify keeping the two rules separate as long we gracefully handle the `noname=yes` `not:name=*` combination that currently occurs on about 100 features.

mtmail replied:
I added logic to look for "tagname" and "not:tagname" and if both have the same value. And if tagname is "name:*" then it's also checked against "not:name". I'd appreciate if you could double-check.

1ec5 commented on data/core.yaml:
```diff
@@ -1867,6 +1867,10 @@ en:
       relation_type:
         message: "{feature} is a relation without a type"
+    mutually_exclusive_tags:
+      title: Mutually Exclusive Tags
+      tip: "Find features having conflicting tags set."
+      message: "{feature} has conflicting tags: {tag1}, {tag2}. Only one should be set."
```
This warning message should be brief; provide more details in a `reference` property:

```suggestion
      title: Contradictory Tags
      tip: "Find features that have contradictory tags."
      message: "{feature} has both {tag1} and {tag2}"
      reference: "The keys {tag1} and {tag2} contradict each other. Only one should be set."
```

mtmail replied:
I updated the titles

1ec5 commented on modules/validations/mutually_exclusive_tags.js:
```diff
@@ -0,0 +1,46 @@
+                    let entity = context.hasEntity(this.entityIds[0]);
+                    return entity ? t.append('issues.' + type + '.message', {
+                        feature: utilDisplayLabel(entity, context.graph()),
+                        tag1: pair[0],
+                        tag2: pair[1]
```
Since these are raw key names, how about wrapping them in `<code>` elements so they look monospaced?

mtmail replied:
I couldn't find any other translation strings using HTML and when supplying values for the `{}` replacements any HTML gets stripped. I added quotes around the name.

1ec5 commented on modules/validations/mutually_exclusive_tags.js:
```diff
@@ -0,0 +1,46 @@
+                        tag1: pair[0],
+                        tag2: pair[1]
+                    }) : '';
+                },
```
In addition to specifying a `reference`, consider supplying a `dynamicFixes` with two fixes that remove either of the contradictory tags.

mtmail replied:
I added fixes now.

mtmail commented:
I added dynamic fixes and better titles, thanks @1ec5.

![image](https://github.com/openstreetmap/iD/assets/3727288/88f6d03b-1fba-4109-bc99-581d58020ffa)

pnorman commented:
> I added dynamic fixes and better titles, thanks @1ec5.
> 
> ![image](https://private-user-images.githubusercontent.com/3727288/289390555-88f6d03b-1fba-4109-bc99-581d58020ffa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzMTUyMzcsIm5iZiI6MTcwMjMxNDkzNywicGF0aCI6Ii8zNzI3Mjg4LzI4OTM5MDU1NS04OGY2ZDAzYi0xZmJhLTQxMDktYmM5OS01ODFkNTgwMjBmZmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTFUMTcxNTM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDVhN2U1ZGU5YTNhNjNhYzE3NmRiZGRkNjIyYjA3NzgzMjM1OTJjZGZkZDZmNmE4MTc4OWNhNzFkYmZjM2YwZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.DK-N_YFI25cQk6VFvq5Ar4OPK62xXPlVxem4RkPJjRE)

There are two quotation marks after name in the text, there should only be one.

mtmail commented:
@pnorman Thank you, fixed now.

1ec5 commented on modules/validations/mutually_exclusive_tags.js:
```diff
@@ -0,0 +1,75 @@
+    const validation = function checkMutuallyExclusiveTags(entity /*, graph */) {
+
+        const pairsFounds = tagPairs.filter((pair) => {
+            return (pair[0] in entity.tags && pair[1] in entity.tags);
```
It isn’t as common, but a negative key can technically be set to the value `no` as well. The double negative would be redundant to the positive tag but wouldn’t contradict it. I guess it’s OK if we still show the warning in this case, but it’s a bit confusing if you know what’s in the raw tags.

mtmail replied:
I added logic now

1ec5 commented on modules/validations/mutually_exclusive_tags.js:
```diff
@@ -0,0 +1,75 @@
+    const tagPairs = [
+        ['noname', 'name'],
+        ['noref', 'ref'],
+        ['noaddress', 'addr:housenumber'],
```
[`nohousenumber`](https://wiki.openstreetmap.org/wiki/Key:nohousenumber) also contradicts `addr:housenumber`.

mtmail replied:
Added to the list now

1ec5 commented on data/core.yaml:
```diff
@@ -1992,6 +1997,9 @@ en:
         annotation: Removed tag.
+      remove_named_tag:
+        title: Remove the '{tag}' tag
+        annotation: Removed '{tag}' tag.
```
Use double quotation marks for consistency with this validator rule’s `reference` and with other validator rules such as `tag_as_not`:

```suggestion
        title: 'Remove the "{tag}" tag'
        annotation: 'Removed "{tag}" tag.'
```

mtmail replied:
good idea

1ec5 commented on modules/validations/mutually_exclusive_tags.js:
```diff
@@ -0,0 +1,97 @@
+        });
+
+        // Addtional:
+        // Check if name and not:name (and similar) are set and both have the same value
```
Thanks for humoring my suggestion. This implementation looks pretty good, but the message for this particular situation should be about the _value_ of `not:name` matching the _value_ of `name`, not that both keys exist. After all, the combination of `name=Mi Casa` and `not:name=Su Casa` is valid. You’re currently checking that correctly, but the warning message is a bit misleading.

Moreover, there are two redundant warnings every time `name` matches `not:name`. Either we should keep the preexisting `not:name` validation rule and special-case this code to skip over `not:name`, or we should remove the preexisting `not:name` rule and clarify the messages similar to the original.

<img src="https://github.com/openstreetmap/iD/assets/1231218/c3714a69-6296-4495-b460-75871da8ced4" width="399" alt="ABC has both not:name and name, Building has the mistaken name “ABC”">

mtmail replied:
I moved logic and tests from the `suspicious_name` validator now. You're right I overlooked that `not:name` can be a list of values, semicolon separated. That's also covered now.

mtmail commented:
Updated screenshots

![image](https://github.com/openstreetmap/iD/assets/3727288/b3c75c15-d585-4b7b-bab6-57eb47fcbaa2)

mtmail commented:
@tyrasd I removed the commented out code, fixed a codespell warning and moved the list of tags to `modules/osm/tags.js`. The branch is rebased with `develop` branch and test pass for me.
